pkg/controller/shared_state_test.go

```diff
@@ -0,0 +1,513 @@
+	azDiskObjs, kubeObjs := splitObjects(objects...)
+	testSharedState := initState(mockclient.NewMockClient(controller), azdiskfakes.NewSimpleClientset(azDiskObjs...), fakev1.NewSimpleClientset(kubeObjs...), objects...)
+
+	return testSharedState
+}
+
+func TestGetNodesForReplica(t *testing.T) {
+	tests := []struct {
+		description string
+		volumes     []string
+		pods        []v1.Pod
+		setupFunc   func(*testing.T, *gomock.Controller) *ReconcileReplica
+		verifyFunc  func(*testing.T, []string, error)
+	}{
+		{
+			description: "[Success] Should not select nodes with no remaining capacity.",
+			volumes:     []string{testPersistentVolume0Name},
+			pods:        []v1.Pod{testPod0},
+			setupFunc: func(t *testing.T, mockCtl *gomock.Controller) *ReconcileReplica {
+				replicaAttachment := testReplicaAzVolumeAttachment
+				now := metav1.Time{Time: metav1.Now().Add(-1000)}
+				replicaAttachment.DeletionTimestamp = &now
+
+				newVolume := testAzVolume0.DeepCopy()
+				newVolume.Status.Detail = &azdiskv1beta2.AzVolumeStatusDetail{
+					VolumeID: testManagedDiskURI0,
+				}
+
+				newNode := testNode0.DeepCopy()
+				newNode.Status.Allocatable[consts.AttachableVolumesField] = resource.MustParse("0")
+
+				controller := NewTestReplicaController(
+					mockCtl,
+					testNamespace,
+					newVolume,
+					&testPersistentVolume0,
+					newNode,
+					&testNode2,
+					&testPod0,
+				)
+
+				mockClients(controller.cachedClient.(*mockclient.MockClient), controller.azClient, controller.kubeClient)
+				return controller
+			},
+			verifyFunc: func(t *testing.T, nodes []string, err error) {
+				require.NoError(t, err)
+				require.Len(t, nodes, 1)
+				require.NotEqual(t, nodes[0], testNode0Name)
+			},
+		},
+		{
+			description: "[Success] Should not create replica attachment on a node that does not match volume's node affinity rule",
+			volumes:     []string{testPersistentVolume0Name},
+			pods:        []v1.Pod{testPod0},
+			setupFunc: func(t *testing.T, mockCtl *gomock.Controller) *ReconcileReplica {
+				replicaAttachment := testReplicaAzVolumeAttachment
+				now := metav1.Time{Time: metav1.Now().Add(-1000)}
+				replicaAttachment.DeletionTimestamp = &now
+
+				newPV := testPersistentVolume0.DeepCopy()
+				newPV.Spec.NodeAffinity = &v1.VolumeNodeAffinity{
+					Required: &v1.NodeSelector{
+						NodeSelectorTerms: []v1.NodeSelectorTerm{{
+							MatchExpressions: []v1.NodeSelectorRequirement{{
+								Key:      consts.TopologyRegionKey,
+								Operator: v1.NodeSelectorOpIn,
+								Values:   []string{"westus2"},
+							}},
+						}},
+					},
+				}
+
+				newVolume := testAzVolume0.DeepCopy()
+				newVolume.Status.Detail = &azdiskv1beta2.AzVolumeStatusDetail{
+					VolumeID: testManagedDiskURI0,
+				}
+
+				newNode := testNode0.DeepCopy()
+				newNode.Labels = map[string]string{consts.TopologyRegionKey: "westus2"}
+
+				controller := NewTestReplicaController(
```

Comment:
Since `getRankedNodesForReplicaAttachments` is implemented by `SharedState`, it seems like we should only need to call `NewTestSharedState` here.

Reply:
Updated!